Lib/pty.py

```diff
@@ -170,3 +175,128 @@ def spawn(argv, master_read=_read, stdin_read=_read):
 
     os.close(master_fd)
     return os.waitpid(pid, 0)[1]
+
+# Author: Soumendra Ganguly.
+SIGWINCH = signal.SIGWINCH
+
+def _login_pty(master_fd, slave_fd):
+    """Given a pty, makes the calling process a session leader, makes the pty slave
+    its controlling terminal, stdin, stdout, and stderr. Closes both pty ends."""
+    # Establish a new session.
+    os.setsid()
+    os.close(master_fd)
+
+    try:
+        fcntl.ioctl(slave_fd, termios.TIOCSCTTY) # Make the pty slave the controlling terminal.
+    except:
+        os.close(slave_fd)
+        raise
+
+    # Slave becomes stdin/stdout/stderr.
+    os.dup2(slave_fd, STDIN_FILENO)
+    os.dup2(slave_fd, STDOUT_FILENO)
+    os.dup2(slave_fd, STDERR_FILENO)
+    if (slave_fd > STDERR_FILENO):
+        os.close(slave_fd)
+
+def _winresz(pty_slave):
+    """Resize window."""
+    w = struct.pack('HHHH', 0, 0, 0, 0)
+    s = fcntl.ioctl(STDIN_FILENO, termios.TIOCGWINSZ, w)
+    fcntl.ioctl(pty_slave, termios.TIOCSWINSZ, s)
+
+def _create_hwinch(pty_slave):
+    """Creates SIGWINCH handler."""
+    def _hwinch(signum, frame):
+        try:
+            _winresz(pty_slave)
+        except:
+            pass
+    return _hwinch
+
+def _cleanup(master_fd, slave_fd, old_hwinch, tty_mode):
+    """Performs cleanup in wspawn."""
+
+    # Close both pty ends.
+    os.close(master_fd)
+    os.close(slave_fd)
+
+    # Restore original tty attributes.
+    if tty_mode != None:
+        tty.tcsetattr(STDIN_FILENO, tty.TCSAFLUSH, tty_mode)
+
+    # Restore original SIGWINCH signal handler.
+    try:
+        signal.signal(SIGWINCH, old_hwinch)
+    except:
+        pass
+
```

Comment:
If the parent loop [ _wcopy ] terminates due to an exception, then that exception is caught by wspawn. Then, we call waitpid on spawned child process so that it does not become a zombie. However, right before calling waitpid, the child process is sent SIGTERM. A second (*) after that, the child process, if alive, receives SIGKILL to ensure that it is no longer running.

(*) Will an adjustable sleep be more useful?

References:
1. https://en.wikipedia.org/wiki/Zombie_process
2. https://man7.org/linux/man-pages/man2/wait.2.html
3. https://man7.org/linux/man-pages/man2/kill.2.html